docs/handling-files.rst

```diff
@@ -0,0 +1,86 @@
+
+.. _file-args:
+
+File Arguments
+-----------------
+
+Click supports working with files with the :class:`File` type. Some notable features are:
+
+*   Support for   ``-`` to mean a special file that refers to stdin/stdout. This is common pattern for POSIX command line utilities, for example wget.
+*   Deals with Unicode and bytes correctly for all versions of Python.
+
+Example:
+
+.. click:example::
+
+    @click.command()
+    @click.argument('input', type=click.File('rb'))
+    @click.argument('output', type=click.File('wb'))
+    def inout(input, output):
+        """Copy contents of INPUT to OUTPUT."""
+        while True:
+            chunk = input.read(1024)
+            if not chunk:
+                break
+            output.write(chunk)
+
+And what it does:
+
+.. click:run::
+
+    with isolated_filesystem():
+        invoke(inout, args=['-', 'hello.txt'], input=['hello'],
+               terminate_input=True)
+        invoke(inout, args=['hello.txt', '-'])
+
+File Path Arguments
+----------------------
+
+For handling paths, the :class:`Path` type is better than a string. Some notable features are:
+
+*   It can run checks such as does the path exist or is the file executable.
+*   It also handles formatting error messages with :func:`format_filename` so any undecodable bytes will be printed nicely.
+
+Example:
+
+.. click:example::
+
+    @click.command()
+    @click.argument('filename', type=click.Path(exists=True))
+    def touch(filename):
+        """Print FILENAME if the file exists."""
+        click.echo(click.format_filename(filename))
+
+And what it does:
+
+.. click:run::
+
+    with isolated_filesystem():
+        with open('hello.txt', 'w') as f:
+            f.write('Hello World!\n')
+        invoke(touch, args=['hello.txt'])
+        println()
+        invoke(touch, args=['missing.txt'])
+
+
+File Opening Behaviors
+-----------------------------
+
+The :class:`File` type attempts to be "intelligent" about when to open a file. Stdin/stdout and files opened for reading will be opened immediately. This will give the user direct feedback when a file cannot be opened. Files opened for writing will only be open on the first IO operation. This is done by automatically wrapping the file in a special wrapper.
+
+File open behavior can be controlled by the boolean kwarg ``lazy``. If a file is opened lazily:
+
+*   A failure at first IO operation will happen by raising an :exc:`FileError`.
+*   It can help minimize resource handling confusion.  If a file is opened in lazy mode, it will receive a ``close_intelligently`` method that can help figure out if the file needs closing or not. This is not needed for parameters, but is necessary for manually prompting. For manual prompts with the :func:`prompt` function you do not know if a stream like stdout was opened (which was already open before) or a real file was opened (that needs closing).
```

Comment:
Do we have a page anywhere where `close_intelligently` is documented? Would be nice to link to it because this does not explain that enough imho.

Reply:
I agree it does not explain enough. It seems like it refers to specific parts of the code but I am not sure which.

Reply:
It refers to `LazyFile.close_intelligently`. Could you add that reference? That should be good for now. `LazyFile` is in utils.